yellow-paper/docs/public-vm/avm-circuit.md

```diff
@@ -51,3 +72,62 @@ An instruction like `ADDRESS` serves as great example because it performs a read
     - memory write
     - flags: `callPointer`, `userMemory = 1` (user memory access)
     - offset: `dstOffset`
+
+## Circuit I/O
+
+### How do "Public Inputs" work in the AVM circuit?
+ZK circuit proof systems generally define some mechanism for "public inputs" for which witness values must be communicated in full to a verifier. The AVM proof system defines its own mechanism for public inputs in which it flags certain trace columns as "public input columns". Any public input columns must be communicated in full to a verifier.
+
+### AVM public inputs structure
+The VM circuit's I/O is defined as the `AvmPublicInputs` structure detailed below:
+```
+AvmSideEffects {
+    newNoteHashes,
+    newNullifiers,
+    newL2ToL1Messages,
+    unencryptedLogs,
+}
+AvmGasResults {
+    l1GasLeft,
+    l2GasLeft,
+    daGasLeft,
+}
+AvmPublicInputs {
+    initialEnvironment: ExecutionEnvironment & {l1GasLeft, l2GasLeft, daGasLeft},
+    calldata: [],
+    sideEffects: AvmSideEffects,
+    storageAccesses,
+    gasResults: {l1GasLeft, l2GasLeft, daGasLeft},
+}
+```
+
+### AVM public input columns
+The `AvmPublicInputs` structure is represented in the VM trace via the following public input columns:
+1. `initialEnvironment` has a dedicated column and is used to initialize the initial call's `ExecutionEnvironment` and `MachineState`
+1. `calldata` has its own dedicated public input column
+1. `sideEffects: AvmSideEffects`
+    - This represents the final `AccruedSubstate` of the initial message call
+    - There is a separate sub-table (columns) for each side-effect vector
+        - Each row in the `newNoteHashes` sub-table contains `{contractAddress, noteHash}`
+        - Each row in the `newNullifiers` sub-table contains `{contractAddress, nullifier}`
+        - Each row in the `newL2ToL1Messages` sub-table contains `{contractAddress, wordIndex, messageWord}`
+            - where a message containing N words takes up N entries with increasing `wordIndex`
+        - Each row in the `unencryptedLogs` sub-table contains `{contractAddress, wordIndex, logWord}`
+            - where a log containing N words takes up N entries with increasing `wordIndex`
+    - Side effects are present in the trace in execution-order
+1. `storageAccesses`
+    - This contains the first and last public storage access for each slot that is accessed during execution
```

Comment:
Is this assuming that public kernel will handle the squashing?

Reply:
I think Zac's the latest thinking is that the VM will actually handle squashing.